Refactor of Scoped CSS

EliasMasche opened the pull request:
Initial PR of the process for Scoped CSS

linkdotnet commented on src/LinkDotNet.Blog.Web/wwwroot/css/components.css:
```diff
@@ -0,0 +1,216 @@
+    width: 125px;
+}
+
+img {
```
This might be "dangerous" as it now targets every "img" whereas it was scoped to the component before. so probably you want something like actions-img or so?

EliasMasche replied:
Understood, gonna check that issue

linkdotnet reviewed (commented):
Currently the bUnit tests are red, as you "renamed" the CSS selectors

EliasMasche commented:
> Currently the bUnit tests are red, as you "renamed" the CSS selectors

So I rename it back to the original or update the bUnit tests?

linkdotnet commented:
> > Currently the bUnit tests are red, as you "renamed" the CSS selectors
> 
> So I rename it back to the original or update the bUnit tests?

I would do the latter

EliasMasche commented:
Removed the empty razor css files and added one that I missed.

An issue that remains is at home, the introduction hasn't show up yet of `IntroductionCard.razor.css` into `home.css` 
and the CSS of `PreviewImage.razor.css` into `components.css` not being functional either
 
[src/LinkDotNet.Blog.Web/wwwroot/css/home.css](https://github.com/linkdotnet/Blog/pull/413/commits/8522dcf669cfb9e63866489d4a1c58dd8e0b450d#diff-ee24ff176e9faf42bac9db93062e2a44244a7dbc256ecfde62d5f4fe35fc7b9c)

![firefox_hlO8DyJBDB](https://github.com/user-attachments/assets/ed8d6cf9-8b2e-4fb1-a5a5-84471344bf0b)

linkdotnet commented:
> Removed the empty razor css files and added one that I missed.
> 
> An issue that remains is at home, the introduction hasn't show up yet of `IntroductionCard.razor.css` into `home.css` and the CSS of `PreviewImage.razor.css` into `components.css` not being functional either
> 
> [src/LinkDotNet.Blog.Web/wwwroot/css/home.css](https://github.com/linkdotnet/Blog/pull/413/commits/8522dcf669cfb9e63866489d4a1c58dd8e0b450d#diff-ee24ff176e9faf42bac9db93062e2a44244a7dbc256ecfde62d5f4fe35fc7b9c)
> 
> ![firefox_hlO8DyJBDB](https://private-user-images.githubusercontent.com/6045426/427307313-ed8d6cf9-8b2e-4fb1-a5a5-84471344bf0b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDMwOTI3NDksIm5iZiI6MTc0MzA5MjQ0OSwicGF0aCI6Ii82MDQ1NDI2LzQyNzMwNzMxMy1lZDhkNmNmOS04YjJlLTRmYjEtYTVhNS04NDQ3MTM0NGJmMGIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMjdUMTYyMDQ5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NWJkMGM5MTRiNGQ4NDRmNzdlODkxMTRlOTAzZWU1Mzg5NjBlMzMxOTVmODVkMDIwMmQ0YjZkODA5YzNmMDIxOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.9ZwTG4jiC9IghAyiL7yNUGmn1rJuNJdChya3wBedbow)

Hmm that might be because of `--profile-background`. Even though it should work. If you open your devtools and check the value of `--profile-background` is there something useful?

linkdotnet commented:
There are some smaller tests that refer to the "old" structure:
```csharp
 var items = cut.FindAll(".profile-keypoints li");
```

EliasMasche commented:
> > Removed the empty razor css files and added one that I missed.
> > An issue that remains is at home, the introduction hasn't show up yet of `IntroductionCard.razor.css` into `home.css` and the CSS of `PreviewImage.razor.css` into `components.css` not being functional either
> > [src/LinkDotNet.Blog.Web/wwwroot/css/home.css](https://github.com/linkdotnet/Blog/pull/413/commits/8522dcf669cfb9e63866489d4a1c58dd8e0b450d#diff-ee24ff176e9faf42bac9db93062e2a44244a7dbc256ecfde62d5f4fe35fc7b9c)
> > ![firefox_hlO8DyJBDB](https://private-user-images.githubusercontent.com/6045426/427307313-ed8d6cf9-8b2e-4fb1-a5a5-84471344bf0b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDMwOTI3NDksIm5iZiI6MTc0MzA5MjQ0OSwicGF0aCI6Ii82MDQ1NDI2LzQyNzMwNzMxMy1lZDhkNmNmOS04YjJlLTRmYjEtYTVhNS04NDQ3MTM0NGJmMGIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMjdUMTYyMDQ5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NWJkMGM5MTRiNGQ4NDRmNzdlODkxMTRlOTAzZWU1Mzg5NjBlMzMxOTVmODVkMDIwMmQ0YjZkODA5YzNmMDIxOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.9ZwTG4jiC9IghAyiL7yNUGmn1rJuNJdChya3wBedbow)
> 
> Hmm that might be because of `--profile-background`. Even though it should work. If you open your devtools and check the value of `--profile-background` is there something useful?

Based on dev tools the profiles images are loaded correctly but don't show up on the page

![vtqOu6cq6C](https://github.com/user-attachments/assets/f92e3dab-7f1b-4f61-afe4-9225e6bbee81)

How it works: `-profile-picture`, `-profile-image`, `.introduction-container`, `.introduction-background`. This is my first time seeing this, and I am super new to dealing with CSS, I feel I am missing something.

Another bug is with PreviewImageRazor for blogposts that takes over the entire space and locks

![firefox_xmNQ8yKUZh](https://github.com/user-attachments/assets/240e21e7-f13f-40fa-9659-b88da817a193)

The css is `img` that I renamed to `comp-img` the code is `PreviewImage.razor` in the HTML code uses the `img` tag how to rename properly or refactor for taking into account the proper CSS code

```html
@if (string.IsNullOrEmpty(PreviewImageUrlFallback))
{
    <img src="@PreviewImageUrl" alt="Preview image blogpost" loading="@LazyLoadTag" decoding="@DecodingTag" />
}
else
{
    <picture>
        <source srcset="@PreviewImageUrl" type="@GetMimeType()"/>
        <img src="@PreviewImageUrlFallback" alt="Preview image blogpost" loading="@LazyLoadTag" decoding="@DecodingTag" />
    </picture>
}
```

EliasMasche commented:
> There are some smaller tests that refer to the "old" structure:
> 
> ```cs
>  var items = cut.FindAll(".profile-keypoints li");
> ```

Do you need me to update them and search for other CSS renames too?

EliasMasche commented:
> 
> Another bug is with PreviewImageRazor for blogposts that takes over the entire space and locks
> 
> ![firefox_xmNQ8yKUZh](https://private-user-images.githubusercontent.com/6045426/427693808-240e21e7-f13f-40fa-9659-b88da817a193.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDMwOTk1MzUsIm5iZiI6MTc0MzA5OTIzNSwicGF0aCI6Ii82MDQ1NDI2LzQyNzY5MzgwOC0yNDBlMjFlNy1mMTNmLTQwZmEtOTY1OS1iODhkYTgxN2ExOTMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMjdUMTgxMzU1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTg5N2IyOTYxZDc3YmViOTBjYjQzODBlYTllYzdkNzczMzRmOTJjOTc5N2ZkNjZkMTdjOGQ3MjgxMWFjOTVmYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.zkSu3wXihpD89F9YsAlh9vx-CWN5Zmu524BGa3mnRSk)
> 
> The css is `img` that I renamed to `comp-img` the code is `PreviewImage.razor` in the HTML code uses the `img` tag how to rename properly or refactor for taking into account the proper CSS code
> 
> ```
> @if (string.IsNullOrEmpty(PreviewImageUrlFallback))
> {
>     <img src="@PreviewImageUrl" alt="Preview image blogpost" loading="@LazyLoadTag" decoding="@DecodingTag" />
> }
> else
> {
>     <picture>
>         <source srcset="@PreviewImageUrl" type="@GetMimeType()"/>
>         <img src="@PreviewImageUrlFallback" alt="Preview image blogpost" loading="@LazyLoadTag" decoding="@DecodingTag" />
>     </picture>
> }
> ```

About this portion of the issue with PreviewImage.razor I did progress adding

```css
.img-container {
	position: relative;
	width: 100%;
	height: 100%;
	overflow: hidden;
}

.comp-img {
	position: absolute;
	top: 0;
	left: 0;
	object-fit: cover;
	height: 100%;
	width: 100%;
}
```

For `PreviewImageUrl` works as before using `<div class="img-container">` and `class="comp-img"`
```html
    <div class="img-container">
        <img class="comp-img" src="@PreviewImageUrl" alt="Preview image blogpost" loading="@LazyLoadTag" decoding="@DecodingTag" /> 
    </div>
```

For `PreviewImageUrlFallback` is not working as in PreviewImageUrl.
```html
    <div class="img-container">
        <picture>
            <source srcset="@PreviewImageUrl" type="@GetMimeType()" />
            <img class="comp-img" src="@PreviewImageUrlFallback" alt="Preview image blogpost" loading="@LazyLoadTag" decoding="@DecodingTag" />
        </picture>
    </div>
```

but for some reason now stopped working haha the PreviewImageUrl too

linkdotnet commented:
> > There are some smaller tests that refer to the "old" structure:
> > ```cs
> >  var items = cut.FindAll(".profile-keypoints li");
> > ```
> 
> Do you need me to update them and search for other CSS renames too?

We have to make the tests work in one way or another. In this regard, the selector has to be aligned to the new "prefix" added.

linkdotnet commented:
I have to checkout the branch later locally. I tried inside GitHub Codespaces and I got some 404 for the images. Not sure if this is because of GitHub Codespace or it showcases the underlying issue on why the images aren't showing up.

linkdotnet commented:
@EliasMasche  I took the opportunity to push to your PR. Hope you don't mind.

EliasMasche commented:
> @EliasMasche I took the opportunity to push to your PR. Hope you don't mind.

Yeah, no problem. the only issue that remains is the home/introduction CSS, right?

linkdotnet commented:
> > @EliasMasche I took the opportunity to push to your PR. Hope you don't mind.
> 
> Yeah, no problem. the only issue that remains is the home/introduction CSS, right?

Technically, yes. So of course everything should work as before, but overall I would like to go a step further, otherwise we just moved the CSS from one file to another.

Ideally:
 1. We can refactor many of the custom CSS classes to bootstrap 5 utility classes
 2. Consolidate the rest. There is much overlap in the content of those classes which can be moved together. 
 3. Have then only one `basic.css` (well and the `icon.css`) to reduce the initial browser load.

linkdotnet commented:
Just a small heads-up, for me the profile picture loads just fine, but the `ShortBlogPost` is a bit "different":

<img width="1511" alt="image" src="https://github.com/user-attachments/assets/68a6f8be-8c08-42e8-b707-8d4678d8f515" />

linkdotnet commented:
The navbar items seem a bit closer than they were before and lost their "bold" font.

linkdotnet commented:
Should also be fixed. Now the interesting part starts: Consolidating :D 
That is where I see the real value in the PR and ticket

EliasMasche commented:
Roger that.

I checked your commits. I pulled into my local branch, and the CSS works now, but I am getting a 404 for the images. 
`Reason: strict-origin-when-cross-origin`

![firefox_obM70UTEbw](https://github.com/user-attachments/assets/90da4e67-bdb4-4054-a2a9-5b0859162d3a)



> > > @EliasMasche I took the opportunity to push to your PR. Hope you don't mind.
> > 
> > 
> > Yeah, no problem. the only issue that remains is the home/introduction CSS, right?
> 
> Technically, yes. So of course everything should work as before, but overall I would like to go a step further, otherwise we just moved the CSS from one file to another.
> 
> Ideally:
> 
>     1. We can refactor many of the custom CSS classes to bootstrap 5 utility classes
> 
>     2. Consolidate the rest. There is much overlap in the content of those classes which can be moved together.
> 
>     3. Have then only one `basic.css` (well and the `icon.css`) to reduce the initial browser load.

The end goal is to put all CSS into `basics.css`, refactor and reduce the custom CSS classes with bootstrap 5 utility classes.

> That is where I see the real value in the PR and ticket

Good thing this time I put the PR as draft as it is more of a proposal and WIP.

linkdotnet commented:
Interesting. By any chance, do you have a chrome based browser to cross-check if it happens there as well?
If so, if you go to "my website": steven-giesel.com - do you see the image? The difference is that the image on my website is hosted on a Blob storage and not as part of the application.

> Good thing this time I put the PR as draft as it is more of a proposal and WIP.

No worries - that is part of the process. Issues/PR's are async in nature and therefore are more likely to have misunderstandings and whatnot.

EliasMasche commented:
Yeah, same issue coming from Edge Browser,

What it dislikes is that it comes from the same application but a different domain or locally triggering CORS if I use it from another website/domain, or blob storage works fine.

EliasMasche commented:
Hello there, sorry ( but may not be sorry) for the disappearance. I got relocated to a new team, a new project with an impossible deadline, so I wasn't able to work.

So now I pushed a new commit, redoing the rename of a class of CSS as it wasn't necessary, I believed they had the same name, I put them all inside `basics.css`

A question about the bootstrap, this sentence: `We can refactor many of the custom CSS classes to Bootstrap 5 utility classes`

Where can I find more information, or is it referring to the internal CSS classes of Bootstrap 5

EliasMasche commented:
I started making some changes to Bootstrap. What is the plan?

Besides, in this commit based on bootstrap, there are two ways to keep the custom class or use an inline. Checking through the repo, this is the only instance, so I guess using an inline CSS is fine 
[`10e8472` (#413)](https://github.com/linkdotnet/Blog/pull/413/commits/10e84726e27e5bc4f725913612227c8f79502ad4)

And I am already triggering some fails in tests because the HTML classes are missing in the razor files. For the moment, I am working on replacing them with Bootstrap but keeping the custom classes in `basics.css` for fallback. I will compare if the appearance is working as well as the old one, and once I am done, I will remove those custom classes.

linkdotnet commented:
> I started making some changes to Bootstrap. What is the plan?

I am not sure if I understand correct? What do you mean by that?

> Besides, in this commit based on bootstrap, there are two ways to keep the custom class or use an inline. Checking through the repo, this is the only instance, so I guess using an inline CSS is fine
[10e8472 (#413](https://github.com/linkdotnet/Blog/pull/413/commits/10e84726e27e5bc4f725913612227c8f79502ad4)

If we keep stuff then still as classes. That is fine. If we see a chance to remove the custom CSS, that would be great, if not - then let's keep it as it is.

>  For the moment, I am working on replacing them with Bootstrap but keeping the custom classes in basics.css for fallback

Why would we need them as fallback? Sure, the tests have to be adopted and that is unfortunately a bit of a shortcoming on the current test-setup that it uses too much the concrete CSS classes. But that is a whole different story :D so for now it might be the best, to change failing ones to the new classes to it still targets the same object.

EliasMasche commented:
> I am not sure if I understand correct? What do you mean by that?

You replied this question in the second, I was asking or trying to ask if I should refactor until it is done or not possible to do it.

> Why would we need them as fallback? Sure, the tests have to be adopted and that is unfortunately a bit of a shortcoming on the current test-setup that it uses too much the concrete CSS classes. But that is a whole different story :D so for now it might be the best, to change failing ones to the new classes to it still targets the same object.

Sorry, what I mean is I still keep the old custom CSS classes for the time being for testing if the new HTML bootstrap CSS classes are working properly. Once I finish, I will do the cleaning. I am keeping a list of the custom CSS classes and commenting as I progress with the refactor

EliasMasche commented:
I did some progress overall, trying to match or mimic the appearance from custom CSS classes with Bootstrap 5 and with that able to reduce LoS in `basic.css`, the `.blog-card` class still not able to achieve progress or at least fully replace it but did at minor places, the rest took time.

There is this new warning:

`Error: /Users/runner/work/Blog/Blog/src/LinkDotNet.Blog.Web/Features/Home/Components/IntroductionCard.razor(42,20): error S1144: Remove the unused private property 'IntroductionStyle'. (https://rules.sonarsource.com/csharp/RSPEC-1144) [/Users/runner/work/Blog/Blog/src/LinkDotNet.Blog.Web/LinkDotNet.Blog.Web.csproj]`

- Is it free to remove safely without any collateral damage?
- Is it fine that I check the rest of the HTML classes based on Bootstrap 4 and update them to 5? I think I found some that are using the semantics of 4

About the tests, how is gonna be the progress going to be?

linkdotnet commented:
Sorry for the late answer - I try to find some time next week for this and give you a proper answer to your questions

EliasMasche commented:
No problemo, as you once wrote, there is time for open-source, sometimes must side step for other priorities.

At least I know still alive the process ^^

linkdotnet commented on src/LinkDotNet.Blog.Web/wwwroot/css/icons.css:
Please keep the icons.css


There is a readme explaining on how to get the icons.css. Therefore this file is "autogenerated" to some extend.

linkdotnet commented:
> About the tests, how is gonna be the progress going to be?

I guess you are not that familiar with **bUnit**. So I can take that over.

It depends on the test, but before they had a css-selector and we still have to provide one so that the test triggers the right HTML element.

EliasMasche commented:
It seems that I closed the PR by accident when I replied because I don't remember doing it.

I pushed new commits based last version, refactoring the `razor.css` files and relocating what is not possible to refactor to Bootstrap 5 utility classes. I feel there could be reduced even more, but I am not experienced with CSS.

A summary of what css it keeps:
- `ShortBlogPost.razor`: the entire CSS got relocated to `basics.css`
- `Profile.razor`: it kept `profile-keypoints` and minor ones
- `IntroductionCard.razor`: I use a mix of inline style for background-image and custom CSS to not trigger CORS while using local assets. There is a minor difference between the old `razor.css` and now, where the background image has been cut up. For example, in the older the background image shows the rocks and mouth of a cat, while this one is cut earlier 
- `TabbedNavigation`: still has its custom CSS
- `SearchInput`: Reduced CSS, a minor difference is that the circle bar is thinner or darker, or both
- `ReadingIndicator`: keeps the custom CSS because it `ReadingIndicator.razor.js` depends on it

linkdotnet commented on src/LinkDotNet.Blog.Web/Features/AboutMe/Components/Skill/SkillTable.razor:
```diff
@@ -5,40 +5,44 @@
+                            if (ShowAdminActions)
                             {
-                                <div draggable="true" @ondrag="@(() => currentDragItem = skill)" style="cursor: grab"
-                                 class="skill-tag">
```
For example this is why some of the tests are failing.
There is a test looking for `.skill-tag` inside tests. 
So we have two options:
 1. Add again `.skill-tag` here without the backing razor.css (kind of just as a marker)
 2. Or replace `skill-tag` with d-inline-block me-2 my-2 px-2 py-1 rounded bg-light`

I would go for the first approach - as it makes the tests more readable

EliasMasche commented:
Roger that, if you find some differences, tell me so I can fix them, I noted some of the differences that I listed them probably are more

linkdotnet commented:
Hey @EliasMasche - sorry for the long pause. Given all the "issues" we discovered with the approach of moving it into central files - is it really worth the effort if it might break left and right?

EliasMasche commented:
Hello there, @linkdotnet is fine, I wasn't able to check much either through the month because I was dealing with the curves that life has been throwing me.

About the question of worth, I think it is fine as is and will not break anymore; only these three issues remain:  

<img width="1768" height="795" alt="471995906-bd591090-b3ca-4f1e-9b92-d2b0051ea8ba" src="https://github.com/user-attachments/assets/4fcd5dae-cc51-4158-ba2b-761c81e9f565" />


About 1. and 2. tried to fix, but weren't able to identify which CSS code or line is breaking the appearance. A detail of issue 2. is that the code is double-drawing the box, so probably rewrite or revamp with a new appearance 


3. I was able to fix or remake the old behavior with spaced navmenu `--bs-navbar-nav-link-padding-x: 1.0rem;` at `:root`

The rest of CSS code got relocated to the central CSS `basics.css` without any other issues so far. 

This week I was gonna resume and verify what I could do or remake the CSS, but I didn't know what or how to progress

EliasMasche commented:
Reverted to the razor.css file. Regarding the merge conflict, I leave it in your hands, or I will check.

EliasMasche commented:
Hey @linkdotnet, I've been checking the errors from the tests and have not been able to find the cause of the failed tests with `SmokeTests.cs` in lines:

https://github.com/linkdotnet/Blog/actions/runs/17752762473/job/50450353066?pr=413#step:6:30 
Method `ShouldBootUpApplication`

https://github.com/linkdotnet/Blog/actions/runs/17752762473/job/50450353066?pr=413#step:6:49
Method `IndexPage_HasTitleFromConfiguration`

Did something from my branch or config get loaded wrong? because that is what it seems.

EliasMasche commented:
All tests passed, all green, it seems ready for the last step when you are ready. On the last commits, I searched for all `styles=` and replaced them with Bootstrap classes or custom CSS.
